irohad/consensus/yac/impl/yac_gate_impl.cpp

```diff
@@ -67,82 +66,84 @@ namespace iroha {
-                          consensus_result_cache_->insert(
-                              std::make_shared<ConsensusResult>(*block));
-                          subscriber.on_next(*block);
-                        }
-                        subscriber.on_completed();
-                      });
-                })
-                // need only the first
-                .first()
-                .retry()
-                .subscribe(
-                    // if load is successful from at least one node
-                    [subscriber](auto block) {
-                      subscriber.on_next(block);
-                      subscriber.on_completed();
-                    },
-                    // if load has failed, no peers provided the block
-                    [this, subscriber](std::exception_ptr) {
-                      log_->error("Cannot load committed block");
-                      subscriber.on_completed();
-                    });
-          });
+              std::shared_ptr<shared_model::interface::Block>>(
+              [this, commit_message](auto subscriber) {
+                const auto hash = getHash(commit_message.votes);
+                if (not hash) {
+                  log_->info("Invalid commit message, hashes are different");
+                  subscriber.on_completed();
+                  return;
+                }
+                // if node has voted for the committed block
+                if (hash == current_block_.first) {
+                  // append signatures of other nodes
+                  this->copySignatures(commit_message);
+                  log_->info("consensus: commit top block: height {}, hash {}",
+                             current_block_.second->height(),
+                             current_block_.second->hash().hex());
+                  subscriber.on_next(current_block_.second);
+                  subscriber.on_completed();
+                  return;
+                }
+                // node has voted for another block - load committed block
+                const auto model_hash =
+                    hash_provider_->toModelHash(hash.value());
+                // iterate over peers who voted for the committed block
+                rxcpp::observable<>::iterate(commit_message.votes)
+                    // allow other peers to apply commit
+                    .flat_map([this, model_hash](auto vote) {
+                      // map vote to block if it can be loaded
+                      return rxcpp::observable<>::create<
+                          std::shared_ptr<shared_model::interface::Block>>(
+                          [this, model_hash, vote](auto subscriber) {
+                            auto block = block_loader_->retrieveBlock(
+                                vote.signature->publicKey(),
+                                shared_model::crypto::Hash(model_hash));
+                            // if load is successful
+                            if (block) {
```

Comment:
_optional:_ could you please explicitly call `has_value`, so it obvious that it's boost::optional?

Reply:
no, boost optional does not have such method

Reply:
only for reference, indeed